### PR TITLE
(WIP) add scopes

### PR DESCRIPTION
**Pull request by BartJanvanAssen**
Still work in progress, but this is adding import maps scopes of #38  to the repo. It's not finished yet, though big part is here. Theres some refinement needed in the web-server. Additionally i think its wise to create some tests around this to expect the same behaviour

### CODE REVIEW

File: src/web-server.js [hunk 1]
```diff
@@ -236,6 +238,60 @@ app.delete("/services/:serviceName", function (req, res) {
     });
 });
 
+app.patch("/scopes", function (req, res) {
```
Comment (joeldenning) [inline src/web-server.js:241]: This is not what was discussed in the Github issue. The plan is to support import map scopes with `PATCH /import-map.json`. We do not need a new api to support it.

File: src/web-server.js [hunk 2]
```diff
@@ -236,6 +238,60 @@ app.delete("/services/:serviceName", function (req, res) {
...
+            .status(500)
+            .send(`Could not write manifest file -- ${ex.toString()}`);
+        });
+    })
+    .catch((err) => {
+      res.status(400).send(err.message);
+    });
+});
+
+app.delete("/scopes/:scopeName", function (req, res) {
```
Comment (joeldenning) [inline src/web-server.js:282]: Hmm can this API be part of PATCH /import-map.json? PATCH requests usually do not delete omitted properties, so maybe we can not. In that case, it might be okay to have this API.

File: src/modify.js [hunk 3]
```diff
@@ -6,25 +6,72 @@ const config = require("./config").config;
...
+exports.modifyMultipleScopes = function (env, newImports) {
+  return modifyMultiple(env, newImports, "scopes");
+};
+
+exports.modifyScope = function (env, serviceName, url, remove) {
+  return modifyKeyValue(env, serviceName, url, remove, "scopes");
+};
+
+function modifyMultiple(env, values, manifestScope) {
+  console.log("modify multi", env, values, manifestScope);
```
Comment (joeldenning) [inline src/modify.js:42]: Please do not push console logs that are used for debugging.

**Comment by BartJanvanAssen** [comment]
What still is left out is the scoping URL validation, which I think requires some setup of jest testing to cover all situations. Any objections to set that up in this pr @joeldenning?

File: src/modify.js [hunk 4]
```diff
@@ -6,25 +6,69 @@ const config = require("./config").config;
...
+};
+
+exports.modifyScope = function (env, serviceName, url, remove) {
+  return modifyKeyValue(env, serviceName, url, remove, "scopes");
+};
+
+function modifyMultiple(env, values, manifestScope) {
+  return modifyLock(env, (json) => {
+    const imports = getScopeFromManifest(json, manifestScope);
+    imports[manifestScope] = values;
```
Comment (joeldenning) [inline src/modify.js:44]: This appears to be wrong in the context of import map scopes. The `imports` key and `scopes` keys are quite different from each other and I think it makes sense to keep their implementations separate.

```json
{
  "imports": {
    "foo": "https://cdn.com/foo.js"
  },
  "scopes": {
    "https://cdn.com/": {
      "bar": "https://cdn.com/bar.js"
    }
  }
}
```

Notice how the values in the `imports` key are strings, but the the values in `scopes` key are objects.

File: src/modify.js [hunk 5]
```diff
@@ -6,25 +6,69 @@ const config = require("./config").config;
...
+    return imports;
+  });
+}
+
+function modifyKeyValue(env, key, value, remove, manifestScope) {
```
Comment (joeldenning) [inline src/modify.js:49]: Similar thing here - imports and scopes are different enough that they should have their own implementation.

**Comment by BartJanvanAssen** [comment]
Updated the pr with more separation of the services and scopes. I'll continue to work on the validation.

File: src/modify.js [hunk 6]
```diff
@@ -88,4 +69,53 @@ exports.modifyMultipleServices = function (env, newImports) {
...
+
+exports.modifyMultipleScopes = function (env, newScopes) {
+  return modifyLock(env, (json) => {
+    const entries = getEntriesForKey(json, "scopes");
+    entries.scopes = newScopes;
+    return entries;
+  });
 };
+
+exports.modifyScope = function (env, scopeSpecifier, scopeMappingObj, remove) {
```
[the remark this answers is not in the document]
Comment (BartJanvanAssen) [reply src/modify.js:109]: Alright, ye i thought lets keep this in for patching a singular scopes. but removing is fine with me

**Comment by BartJanvanAssen** [comment]
@joeldenning i see travis is doing some eslint stuff over here, can i extend the eslint in this project to also check in the precommit hook?

File: src/modify.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (joeldenning) [reply src/modify.js]: I prefer only adding code when it's being used.

File: src/verify-valid-url.js [hunk 7]
```diff
@@ -23,4 +22,21 @@ exports.verifyValidUrl = async function (req, url) {
       );
     }
   }
+}
+
+exports.verifyValidUrl = verifyValidUrl;
+
+exports.findUrlsToValidateInScopes = function (scopes) {
```
Comment (joeldenning) [inline src/verify-valid-url.js:29]: I think this function would be simpler with for loops, or even with a reduce. I won't block approval on this, but thought I'd share an implementation that I thought made sense

```js
exports.findUrlsToValidateInScopes = function (scopes) {
  const result = []

  for (let scopeKey in scopes) {
    const scope = scopes[scopeKey]
    for (let specifier in scope) {
      const address = scope[specifier]
      // create url to validate
    }
  }
}
```

File: src/verify-valid-url.js [hunk 8]
```diff
@@ -23,4 +22,21 @@ exports.verifyValidUrl = async function (req, url) {
...
+  const toValidateUrls = [];
+  Object.keys(scopes).map((scope) => {
+    const scopeOverrides = Object.entries(scopes[scope]);
+    scopeOverrides.map(([specifier, address]) => {
+      // we cannot validate relative urls
+      if (scope.includes("http")) {
```
Comment (joeldenning) [inline src/verify-valid-url.js:35]:
```suggestion
      if (scope.startsWith("http://") || scope.startsWith('https://')) {
```

File: src/verify-valid-url.js [hunk 9]
```diff
@@ -23,4 +22,21 @@ exports.verifyValidUrl = async function (req, url) {
...
+        toValidateUrls.push(new URL(address, scope).toString());
```
Comment (joeldenning) [inline src/verify-valid-url.js:36]: Relying on `toString()` in javascript is always a bit scary - is better to use the `href` property on URLs:

https://developer.mozilla.org/en-US/docs/Web/API/URL/href

```suggestion
        toValidateUrls.push(new URL(address, scope).href);
```

File: src/verify-valid-url.js (hunk 8 above)
Comment (joeldenning) [reply src/verify-valid-url.js:35]: We can verify addresses that start with http, as well.

```js
if (canVerify(scope)) {
  toValidateUrls.push(new URL(address, scope).href);
} else if (canVerify(address)) {
  toValidateUrls.push(address)
}

function canVerify(url) {
  return url.startsWith('http://') || url.startsWith('https://')
}
```

File: test/import-map.test.js [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+    password: "password",
+    locations: {
+      prod: "memory://prod",
+    },
+  });
+});
+
+describe(`/import-map.json`, () => {
+  it(`does not return anything when it's not setup yet.`, async () => {
+    const healthResponse = await request(app)
```
Comment (joeldenning) [inline test/import-map.test.js:17]: This is not a response to the health endpoint:

```suggestion
    const response = await request(app)
```

File: test/import-map.test.js [hunk 11]
```diff
@@ -0,0 +1,106 @@
+const request = require("supertest");
+const { app, setConfig } = require("../src/web-server");
+
+beforeAll(() => {
+  setConfig({
+    manifestFormat: "importmap",
+    username: "username",
+    password: "password",
```
Comment (joeldenning) [inline test/import-map.test.js:8]: Why have a username and password for tests? I only did so for the health endpoint because I was specifically testing authentication related behavior. However, since these tests don't test auth, they don't need it.

```suggestion
```

File: test/import-map.test.js [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+
+describe(`/import-map.json`, () => {
+  it(`does not return anything when it's not setup yet.`, async () => {
+    const healthResponse = await request(app)
+      .get("/import-map.json")
+      .expect(200)
+      .expect("Content-Type", /json/);
+
+    // we did not setup yet
+    expect(healthResponse.body.message).toBe(undefined);
```
Comment (joeldenning) [inline test/import-map.test.js:23]: Shouldn't this return an empty import map? Since if the io-method returns an empty string, we create an empty import map.

File: test/import-map.test.js [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+      })
+      .expect(200)
+      .expect("Content-Type", /json/);
+
+    // we did not setup yet
+    expect(healthResponse.body).toMatchObject({
+      a: "/a-1-updated.mjs",
+      b: "/b-1.mjs",
+      c: "/c-1.mjs",
+    });
```
Comment (joeldenning) [inline test/import-map.test.js:90]: This part of the test seems to rely on test execution order, which is not good. We should not require the tests to execute in a certain order.

```suggestion
    expect(healthResponse.body).toMatchObject({
      a: "/a-1-updated.mjs",
    });
```

File: test/import-map.test.js [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+      imports: {
+        a: "/a-1.mjs",
+        b: "/b-1.mjs",
+        c: "/c-1.mjs",
+      },
+      scopes: {},
+    });
+  });
+
+  it(`The data is still in there`, async () => {
```
Comment (joeldenning) [inline test/import-map.test.js:54]: This test relies on test execution order, which makes tests very brittle. We should not require the test on line 26 to execute before this test in order for this test to pass. Each test should be independent.

File: test/import-map.test.js [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+      .set("accept", "json")
+      .expect(200)
+      .expect("Content-Type", /json/);
+
+    // we did not setup yet
+    expect(healthResponse.body).toMatchObject({
+      a: "/a-1-updated.mjs",
+      c: "/c-1.mjs",
+    });
+  });
```
Comment (joeldenning) [inline test/import-map.test.js:105]: To avoid relying on test execution order, we should just verify that service b is no longer in the import map.

```suggestion
    expect(healthResponse.body.b).toBe(undefined);
```

File: test/scope-map.test.js [hunk 16]
```diff
@@ -0,0 +1,92 @@
+const request = require("supertest");
+const { app, setConfig } = require("../src/web-server");
+
+beforeAll(() => {
+  setConfig({
+    manifestFormat: "importmap",
+    username: "username",
```
Comment (joeldenning) [inline test/scope-map.test.js:7]: Remove this username and password

File: test/scope-map.test.js [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+
+describe(`/import-map.json - Scopes`, () => {
+  it(`does not return anything when it's not setup yet.`, async () => {
+    const healthResponse = await request(app)
+      .get("/import-map.json")
+      .expect(200)
+      .expect("Content-Type", /json/);
+
+    // we did not setup yet
+    expect(healthResponse.body.message).toBe(undefined);
```
Comment (joeldenning) [inline test/scope-map.test.js:23]: I would expect this to be an empty import map

File: test/scope-map.test.js [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+    password: "password",
+    locations: {
+      prod: "memory://prod",
+    },
+  });
+});
+
+describe(`/import-map.json - Scopes`, () => {
+  it(`does not return anything when it's not setup yet.`, async () => {
+    const healthResponse = await request(app)
```
Comment (joeldenning) [inline test/scope-map.test.js:17]:
```suggestion
    const response = await request(app)
```

File: test/scope-map.test.js [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+          },
+          "https://cdn.com/scope2/scope3/": {
+            b: "https://cdn.com/b-3.mjs",
+          },
+        },
+      })
+      .expect(200)
+      .expect("Content-Type", /json/);
+
+    // we did not setup yet
```
Comment (joeldenning) [inline test/scope-map.test.js:51]: This comment is a copy paste error

File: test/scope-map.test.js [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+          a: "https://cdn.com/a-2.mjs",
+        },
+        "https://cdn.com/scope2/scope3/": {
+          b: "https://cdn.com/b-3.mjs",
+        },
+      },
+    });
+  });
+
+  it(`The data is still in there`, async () => {
```
Comment (joeldenning) [inline test/scope-map.test.js:69]: This test should be deleted, since it relies on test execution order

File: test/scope-validation.test.js [hunk 21]
```diff
@@ -0,0 +1,69 @@
+const { findUrlsToValidateInScopes } = require("../src/verify-valid-url");
+
+describe(`/import-map.json - Scopes`, () => {
+  // example https://github.com/WICG/import-maps#scoping-examples
+  it(`does found all urls that have to be validated.`, async () => {
```
Comment (joeldenning) [inline test/scope-validation.test.js:5]:
```suggestion
  it(`finds all urls that have to be validated.`, async () => {
```

File: test/import-map.test.js (hunk 12 above)
Comment (BartJanvanAssen) [reply test/import-map.test.js:23]: Nope its something i’ve noticed twice setting up the project, would be nice to solve 😊

File: src/verify-valid-url.js (hunk 7 above)
Comment (BartJanvanAssen) [reply src/verify-valid-url.js:29]: I think that makes total sense, let me refactor :)

File: test/import-map.test.js (hunk 12 above)
Comment (BartJanvanAssen) [reply test/import-map.test.js:23]: Would you, @joeldenning, want to fix that in this pr, or create a new pr for that one?

File: src/modify.js [hunk 22]
```diff
@@ -88,4 +69,61 @@ exports.modifyMultipleServices = function (env, newImports) {
...
   });
+}
+
+/*
+ * Services
+ */
+exports.modifyMultipleServices = function (env, newImports) {
+  return modifyLock(env, (json) => {
+    const imports = getMapFromManifest(json);
+    if (imports) {
```
[the remark this answers is not in the document]
Comment (BartJanvanAssen) [reply src/modify.js:80]: I thought i could reproduce it, though couldn't.

[the remark this answers is not in the document]
Comment (joeldenning) [reply src/modify.js:80]: Please remove this then

File: test/import-map.test.js [hunk 23]
```diff
@@ -0,0 +1,82 @@
...
+        imports: {
+          a: "/a-1.mjs",
+          b: "/b-1.mjs",
+          c: "/c-1.mjs",
+        },
+      })
+      .expect(200)
+      .expect("Content-Type", /json/);
+
+    // we did not setup yet
```
Comment (joeldenning) [inline test/import-map.test.js:41]: This comment should be removed, as it's a copy pasted message that doesn't make sense here.

```suggestion
```

File: test/import-map.test.js [hunk 24]
```diff
@@ -0,0 +1,82 @@
...
+      })
+      .set("accept", "json")
+      .send({
+        service: "a",
+        url: "/a-1-updated.mjs",
+      })
+      .expect(200)
+      .expect("Content-Type", /json/);
+
+    // we did not setup yet
```
Comment (joeldenning) [inline test/import-map.test.js:66]:
```suggestion
```

File: test/scope-map.test.js [hunk 25]
```diff
@@ -0,0 +1,99 @@
...
+        scopes: {
+          "https://cdn.com/scope2/": {
+            a: "https://cdn.com/a-2.mjs",
+          },
+          "https://cdn.com/scope2/scope3/": {
+            b: "https://cdn.com/b-3.mjs",
+          },
+        },
+      });
+    // .expect(200)
```
Comment (joeldenning) [inline test/scope-map.test.js:40]: Why is this commented?

File: test/scope-map.test.js [hunk 26]
```diff
@@ -0,0 +1,99 @@
...
+    // .expect("Content-Type", /json/);
+
+    // we did not setup yet
```
Comment (joeldenning) [inline test/scope-map.test.js:43]:
```suggestion
```

File: src/web-server.js [hunk 27]
```diff
@@ -118,54 +120,111 @@ app.patch("/import-map.json", (req, res) => {
...
+        .send(`Invalid import map in request body -- imports is not an object`);
+    }
 
-  if (req.body.scopes) {
-    res
-      .status(400)
-      .send("import-map-deployer does not support import map scopes");
-    return;
-  }
+    if (Object.keys(req.body.imports).length === 0) {
```
[the remark this answers is not in the document]
Comment (BartJanvanAssen) [reply src/web-server.js:130]: Would have my preference, firing an empty import/scope object isn't hurting anything.

File: src/web-server.js [hunk 28]
```diff
@@ -118,54 +120,111 @@ app.patch("/import-map.json", (req, res) => {
...
+      }
     }
   }
 
-  const importUrls = Object.values(req.body.imports);
+  let validImportUrlPromises = Promise.resolve();
+  if (req.body.imports) {
+    // Confirm the imports are working
+    const importUrls = Object.values(req.body.imports);
+    const unsafeUrls = importUrls.map(checkUrlUnsafe).filter(Boolean);
```
[the remark this answers is not in the document]
Comment (BartJanvanAssen) [reply src/web-server.js:196]: ye good one, would be good to add a test for it as well, though maybe add that in another pr.. For testing this i need to mock fetch first, because the urls resolve checker will be the first that hits, i would likely need something like: https://www.npmjs.com/package/jest-fetch-mock. (also reversing the check would be desirable i think.. unsafe urls shouldn't be needed to resolve checked)

File: src/web-server.js (hunk 27 above)
[the remark this answers is not in the document]
Comment (joeldenning) [reply src/web-server.js]: http PATCH allows you to omit properties that you are not changing. I would prefer following that convention.

File: src/modify.js [hunk 29]
```diff
@@ -88,4 +69,45 @@ exports.modifyMultipleServices = function (env, newImports) {
...
   });
+}
+
+/*
+ * Services
+ */
+exports.modifyMultipleServices = function (env, newImports) {
+  return modifyLock(env, (json) => {
+    const imports = getMapFromManifest(json);
+    Object.assign(imports, newImports);
```
Comment (joeldenning) [inline src/modify.js:80]: It is now possible for `newImports` to be undefined, since you can call PATCH /import-map.json without providing the `imports` property.

`Object.assign(imports, undefined)` does not throw any errors or cause unexpected behavior. However, it unnecessarily seizes the modify lock and writes the object without the object having changed. Additionally, leaving the code like this makes it more likely for us to introduce a bug later because the code seems to assume that newImports is defined. Instead, we should change this to be similar to what modifyMultipleScopes is:

```js
 exports.modifyMultipleServices = function (env, newImports) {
  if (newImports) {
    return modifyLock(env, json => {
      const imports = getMapFromManifest(json);
      Object.assign(imports, newImports)
      return json
    }) 
  } else {
    return Promise.resolve()
  }
}
```

File: src/web-server.js [hunk 30]
```diff
@@ -118,54 +120,102 @@ app.patch("/import-map.json", (req, res) => {
...
+  }
 
-  Promise.all(validImportUrlPromises)
+  return Promise.all([validImportUrlPromises, validScopeUrlPromises])
     .then(() => {
-      modify
-        .modifyMultipleServices(env, req.body.imports)
+      Promise.all([
+        modify.modifyMultipleServices(env, req.body.imports),
+        modify.modifyMultipleScopes(env, req.body.scopes),
```
Comment (joeldenning) [inline src/web-server.js:212]: There is a problem with this code - it modifies the `imports` and `scopes` as separate "modify locks." This means that the work happens sequentially, and we write the import map to the target destination twice. This has the chance for bugs, since updating either the imports or scopes first without simultaneously updating the other could result in the deployed import map having a bad state. That bad state would usually exist only for less than a second, but even that period of time is problematic for a production environment. Additionally, if the first modification succeeds but the second one fails, the import map would remain in the bad state indefinitely. Updating the import map should be atomic - should all succeed or fail together.

In other words, this code will cause a production environment to have the wrong import map.

To fix, we should modify both the services and the scopes all in the same modifyLock. This code will need to be refactored to do so.

File: test/scope-map.test.js [hunk 31]
```diff
@@ -0,0 +1,105 @@
...
+  it(`does return empty import-map when it's not setup yet.`, async () => {
+    const response = await request(app)
+      .get("/import-map.json")
+      .expect(200)
+      .expect("Content-Type", /json/);
+
+    expect(response.body).toMatchObject({ imports: {}, scopes: {} });
+  });
+
+  it(`does accept an scopes only patch`, async () => {
```
Comment (joeldenning) [inline test/scope-map.test.js:31]:
```suggestion
  it(`does accept a scopes-only patch`, async () => {
```

File: test/scope-map.test.js [hunk 32]
```diff
@@ -0,0 +1,105 @@
...
+    },
+  });
+});
+
+beforeEach(() => {
+  // assure we have a clean import map every test
+  resetMemoryManifest();
+});
+
+describe(`/import-map.json - Scopes`, () => {
```
Comment (joeldenning) [inline test/scope-map.test.js:21]: Let's add a test for an imports-only patch.

File: test/scope-validation.test.js [hunk 33]
```diff
@@ -0,0 +1,69 @@
...
+        },
+      },
+    };
+
+    const toBeVerifiedUrls = findUrlsToValidateInScopes(mockImportMap.scopes);
+
+    expect(toBeVerifiedUrls).toEqual(["https://cdn.com/a-2.mjs"]);
+  });
+
+  it(`does found all urls that have to be validated.`, async () => {
```
Comment (joeldenning) [inline test/scope-validation.test.js:51]:
```suggestion
  it(`does find all urls that have to be validated.`, async () => {
```
